lib/switchtec.c

```diff
@@ -652,33 +698,306 @@ int switchtec_hard_reset(struct switchtec_dev *dev)
 			     NULL, 0);
 }
 
-static int log_a_to_file(struct switchtec_dev *dev, int sub_cmd_id, int fd)
+/**
+ * @brief Free log definition data
+ * @param[in] defs - log definition data to free
+ */
+static void free_log_defs(struct log_defs *defs)
+{
+	int i, j;
+
+	if (!defs->module_defs)
+		return;
+
+	for (i = 0; i < defs->num_alloc; i++) {
+		for (j = 0; j < defs->module_defs[i].num_entries; j++)
+			free(defs->module_defs[i].entries[j]);
+
+		free(defs->module_defs[i].entries);
+	}
+
+	free(defs->module_defs);
+}
+
+/**
+ * @brief Allocate / reallocate log definition data
+ * @param[in] defs 	  - log definition data
+ * @param[in] num_modules - number of modules to allocate for
+ * @return 0 on success, negative value on failure
+ */
+static int realloc_log_defs(struct log_defs *defs, int num_modules)
+{
+	int i;
+
+	defs->module_defs = realloc(defs->module_defs,
+				    (num_modules *
+				     sizeof(struct module_log_defs)));
+	if (!defs->module_defs) {
+		free_log_defs(defs);
+		return -1;
+	}
+
+	for (i = defs->num_alloc; i < num_modules; i++)
+		memset(&defs->module_defs[i], 0,
+		       sizeof(struct module_log_defs));
+
+	defs->num_alloc = num_modules;
+
+	return 0;
+}
+
+/**
+ * @brief Read a log definition file and store the definitions
+ * @param[in] log_def_file - log definition file
+ * @param[out] defs 	   - log definitions
+ * @return 0 on success, negative value on failure
+ */
+static int read_log_defs(FILE *log_def_file, struct log_defs *defs)
 {
 	int ret;
+	char line[LOG_ENTRY_LEN_MAX];
+	char mod_name[FW_MOD_NAME_LEN_MAX];
+	int mod_id;
+	int num_entries;
+	struct module_log_defs *mod_defs;
+	int i;
+
+	/* allocate some log definition entries */
+	ret = realloc_log_defs(defs, INIT_ALLOC_LOG_MODULES);
+	if (ret < 0)
+		return ret;
+
+	while (fgets(line, sizeof(line), log_def_file)) {
+		/*
+		 * Parse the module name, module ID, and number of module log
+		 * entries
+		 */
+		if (sscanf(line, "%31s %i %i",
+			   mod_name, &mod_id, &num_entries) != 3)
+			continue;
+
+		/* reallocate more log definition entries if needed */
+		if (mod_id > defs->num_alloc) {
+			ret = realloc_log_defs(defs, mod_id * 2);
+			if (ret < 0)
+				return ret;
+		}
+
+		mod_defs = &defs->module_defs[mod_id];
+
+		/*
+		 * Skip this module if it has already been done. This can happen
+		 * if the module is duplicated in the log definition file.
+		 */
+		if (strlen(mod_defs->mod_name) != 0) {
+			for (i = 0; i < num_entries; i++) {
+				if (!fgets(line, sizeof(line),
+					  log_def_file))
+					break;
+			}
+			continue;
+		}
```

Comment:
So if an ID exists twice, the subsequent one is ignored? What if they are different? I'd suggest it's better to just error out instead to help ensure you don't have confusing log-def files with multiple instances of the same module.

Reply:
Yes, the subsequent module and its log entries will be ignored. If a module is duplicated then all the log entries will be identical as well. This was the case for our log definition files as there was a FW bug that caused some of the modules to be duplicated. The bug was fixed a couple days ago, but we want to allow parsing to still work when using the older definition files.